File: python/swiftbip39/bip39.py
from __future__ import annotations

import hashlib
import secrets
import unicodedata
from importlib import resources


class BIP39Error(ValueError):
    pass


_VALID_ENTROPY_BYTES = {16, 20, 24, 28, 32}
_VALID_WORD_COUNTS = {12, 15, 18, 21, 24}


def _wordlist() -> list[str]:
    text = resources.files(__package__).joinpath("english.txt").read_text(encoding="utf-8")
    words = text.splitlines()
    if len(words) != 2048:
        raise BIP39Error("BIP-39 English word list must contain 2048 words")
    return words


WORDS = _wordlist()
INDEX_BY_WORD = {word: index for index, word in enumerate(WORDS)}


def generate_mnemonic(words: int = 12) -> str:
    strength = _words_to_strength(words)
    return entropy_to_mnemonic(secrets.token_bytes(strength // 8))


def entropy_to_mnemonic(entropy: bytes | bytearray | str) -> str:
    if isinstance(entropy, str):
        entropy_bytes = bytes.fromhex(entropy.removeprefix("0x"))
    else:
        entropy_bytes = bytes(entropy)

    if len(entropy_bytes) not in _VALID_ENTROPY_BYTES:
        raise BIP39Error(f"Invalid entropy byte count: {len(entropy_bytes)}")

    entropy_bits = _bytes_to_bits(entropy_bytes)
    checksum_length = len(entropy_bytes) * 8 // 32
    checksum_bits = _bytes_to_bits(hashlib.sha256(entropy_bytes).digest())[:checksum_length]
    bits = entropy_bits + checksum_bits
    return " ".join(WORDS[int(bits[index:index + 11], 2)] for index in range(0, len(bits), 11))


def parse_mnemonic(phrase: str) -> str:
    words = _normalize_words(phrase)
    if len(words) not in _VALID_WORD_COUNTS:
        raise BIP39Error(f"Invalid word count: {len(words)}")

    try:
        bits = "".join(f"{INDEX_BY_WORD[word]:011b}" for word in words)
    except KeyError as error:
        raise BIP39Error(f"Unknown word: {error.args[0]}") from error

    checksum_length = len(bits) // 33
    entropy_length = len(bits) - checksum_length
    entropy_bits = bits[:entropy_length]
    checksum_bits = bits[entropy_length:]
    entropy = _bits_to_bytes(entropy_bits)
    expected = _bytes_to_bits(hashlib.sha256(entropy).digest())[:checksum_length]

    if checksum_bits != expected:
        raise BIP39Error("Invalid checksum")

    return " ".join(words)


def validate_mnemonic(phrase: str) -> bool:
    try:
        parse_mnemonic(phrase)
        return True
    except BIP39Error:
        return False


def mnemonic_to_seed(phrase: str, passphrase: str = "") -> bytes:
    mnemonic = unicodedata.normalize("NFKD", parse_mnemonic(phrase))
    salt = unicodedata.normalize("NFKD", "mnemonic" + passphrase)
    return hashlib.pbkdf2_hmac("sha512", mnemonic.encode(), salt.encode(), 2048, 64)


def mnemonic_to_seed_hex(phrase: str, passphrase: str = "") -> str:
    return mnemonic_to_seed(phrase, passphrase).hex()


def _words_to_strength(words: int) -> int:
    if words not in _VALID_WORD_COUNTS:
        raise BIP39Error(f"Invalid word count: {words}")
    return words // 3 * 32


def _normalize_words(phrase: str) -> list[str]:
    return unicodedata.normalize("NFKD", phrase).split()


def _bytes_to_bits(data: bytes) -> str:
    return "".join(f"{byte:08b}" for byte in data)


def _bits_to_bytes(bits: str) -> bytes:
    return bytes(int(bits[index:index + 8], 2) for index in range(0, len(bits), 8))

File: go/bip39/bip39.go
package bip39

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"embed"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
)

//go:embed english.txt
var files embed.FS

var (
	words       []string
	indexByWord map[string]int
)

func init() {
	data, err := files.ReadFile("english.txt")
	if err != nil {
		panic(err)
	}
	words = strings.Fields(string(data))
	if len(words) != 2048 {
		panic("BIP-39 English word list must contain 2048 words")
	}
	indexByWord = make(map[string]int, len(words))
	for i, word := range words {
		indexByWord[word] = i
	}
}

func Generate(wordsCount int) (string, error) {
	strength, err := wordsToStrength(wordsCount)
	if err != nil {
		return "", err
	}
	entropy := make([]byte, strength/8)
	if _, err := rand.Read(entropy); err != nil {
		return "", err
	}
	return EntropyToMnemonic(entropy)
}

func EntropyHexToMnemonic(value string) (string, error) {
	entropy, err := hex.DecodeString(strings.TrimPrefix(value, "0x"))
	if err != nil {
		return "", err
	}
	return EntropyToMnemonic(entropy)
}

func EntropyToMnemonic(entropy []byte) (string, error) {
	if !validEntropyBytes(len(entropy)) {
		return "", fmt.Errorf("invalid entropy byte count: %d", len(entropy))
	}
	checksum := sha256.Sum256(entropy)
	checksumBits := len(entropy) * 8 / 32
	totalBits := len(entropy)*8 + checksumBits
	out := make([]string, 0, totalBits/11)

	for wordIndex := 0; wordIndex < totalBits/11; wordIndex++ {
		index := 0
		for offset := 0; offset < 11; offset++ {
			bitIndex := wordIndex*11 + offset
			var bit int
			if bitIndex < len(entropy)*8 {
				bit = bitAt(entropy, bitIndex)
			} else {
				bit = bitAt(checksum[:], bitIndex-len(entropy)*8)
			}
			index = (index << 1) | bit
		}
		out = append(out, words[index])
	}

	return strings.Join(out, " "), nil
}

func Parse(phrase string) (string, error) {
	parts := strings.Fields(phrase)
	if !validWordCount(len(parts)) {
		return "", fmt.Errorf("invalid word count: %d", len(parts))
	}

	totalBits := len(parts) * 11
	checksumBits := totalBits / 33
	entropyBits := totalBits - checksumBits
	entropy := make([]byte, entropyBits/8)

	for bitIndex := 0; bitIndex < entropyBits; bitIndex++ {
		bit, err := phraseBit(parts, bitIndex)
		if err != nil {
			return "", err
		}
		setBit(entropy, bitIndex, bit)
	}

	checksum := sha256.Sum256(entropy)
	for offset := 0; offset < checksumBits; offset++ {
		expected := bitAt(checksum[:], offset)
		actual, err := phraseBit(parts, entropyBits+offset)
		if err != nil {
			return "", err
		}
		if actual != expected {
			return "", errors.New("invalid checksum")
		}
	}

	return strings.Join(parts, " "), nil
}

func Validate(phrase string) bool {
	_, err := Parse(phrase)
	return err == nil
}

func Seed(phrase string, passphrase string) ([]byte, error) {
	mnemonic, err := Parse(phrase)
	if err != nil {
		return nil, err
	}
	return pbkdf2SHA512([]byte(mnemonic), []byte("mnemonic"+passphrase), 2048, 64), nil
}

func SeedHex(phrase string, passphrase string) (string, error) {
	seed, err := Seed(phrase, passphrase)
	if err != nil {
		return "", err
	}
	return hex.EncodeToString(seed), nil
}

func phraseBit(parts []string, bitIndex int) (int, error) {
	wordIndex := bitIndex / 11
	offset := bitIndex % 11
	index, ok := indexByWord[parts[wordIndex]]
	if !ok {
		return 0, fmt.Errorf("unknown word: %s", parts[wordIndex])
	}
	return (index >> (10 - offset)) & 1, nil
}

func validEntropyBytes(count int) bool {
	return count == 16 || count == 20 || count == 24 || count == 28 || count == 32
}

func validWordCount(count int) bool {
	return count == 12 || count == 15 || count == 18 || count == 21 || count == 24
}

func wordsToStrength(count int) (int, error) {
	if !validWordCount(count) {
		return 0, fmt.Errorf("invalid word count: %d", count)
	}
	return count / 3 * 32, nil
}

func bitAt(data []byte, bitIndex int) int {
	return int((data[bitIndex/8] >> (7 - bitIndex%8)) & 1)
}

func setBit(data []byte, bitIndex int, value int) {
	if value == 1 {
		data[bitIndex/8] |= 1 << (7 - bitIndex%8)
	}
}

func pbkdf2SHA512(password []byte, salt []byte, iterations int, keyLen int) []byte {
	hashLen := 64
	blocks := (keyLen + hashLen - 1) / hashLen
	out := make([]byte, 0, blocks*hashLen)

	for block := 1; block <= blocks; block++ {
		counter := make([]byte, 4)
		binary.BigEndian.PutUint32(counter, uint32(block))
		u := hmacSHA512(password, append(append([]byte{}, salt...), counter...))
		t := append([]byte{}, u...)
		for i := 2; i <= iterations; i++ {
			u = hmacSHA512(password, u)
			for j := range t {
				t[j] ^= u[j]
			}
		}
		out = append(out, t...)
	}

	return out[:keyLen]
}

func hmacSHA512(key []byte, data []byte) []byte {
	mac := hmac.New(sha512.New, key)
	mac.Write(data)
	return mac.Sum(nil)
}
